### PR TITLE
Add methods to load and save the models as bytearrays

### PR DESCRIPTION
**Pull request by madskonradsen**

### CODE REVIEW

**Comment by mdabros** [comment]
Hi @madskonradsen. Thanks for adding this PR. Definitely useful supporting load/save via byte arrays. I will try to review in the next couple of days.

File: src/XGBoostSharp.Tests/XGBClassifierTest.cs [hunk 1]
```diff
@@ -69,6 +70,25 @@ public void XGBClassifierTest_SaveAndLoad()
         TestUtils.AssertAreEqual(expected, actual);
     }
 
+    [TestMethod]
+    public void XGBClassifierTest_SaveAndLoadRaw()
```
Comment (mdabros) [inline src/XGBoostSharp.Tests/XGBClassifierTest.cs:74]: @madskonradsen Seems the test fails on CI:
https://github.com/mdabros/XGBoostSharp/actions/runs/11380712392/job/31695853717?pr=32
```
 Failed XGBClassifierTest_SaveAndLoadRaw [29 ms]
  Error Message:
   Assert.AreEqual failed. Expected a difference no greater than <0.0001> between expected value <0.2789577> and actual value <0>. 
  Stack Trace:
     at XGBoostSharp.Test.TestUtils.AssertAreEqual(Single[] expected, Single[] actual) in /home/runner/work/XGBoostSharp/XGBoostSharp/src/XGBoostSharp.Tests/TestUtils.cs:line 26
   at XGBoostSharp.Test.XGBRegressorTest.XGBClassifierTest_SaveAndLoadRaw() in /home/runner/work/XGBoostSharp/XGBoostSharp/src/XGBoostSharp.Tests/XGBRegressorTest.cs:line 72
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
```

Can you take a look at this before I start reviewing?

Comment (madskonradsen) [reply src/XGBoostSharp.Tests/XGBClassifierTest.cs]: Sure, that was a just a simple mixup :) Nice that CI has been enabled.

**Comment by mdabros** [comment]
@madskonradsen I have added a few comments, besides that it looks good to me 👍 .

**Comment by madskonradsen** [comment]
I can’t see any comments on this PR? :)

**Review by mdabros — COMMENTED** [review]
@madskonradsen Seems I did not "publish" my review comments. I hope you can see them now :-).

File: src/XGBoostSharp/lib/ModelFormat.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿namespace XGBoostSharp.lib;
+
+public enum ModelFormat
```
[the remark this answers is not in the document]
Comment (madskonradsen) [reply src/XGBoostSharp/lib/ModelFormat.cs]: Fair point :)

File: src/XGBoostSharp/lib/Booster.cs [hunk 3]
```diff
@@ -39,6 +40,40 @@ public Booster(string fileName)
         m_safeBoosterHandle = new SafeBoosterHandle(handle);
     }
 
+    public Booster(byte[] bytes)
```
[the remark this answers is not in the document]
Comment (madskonradsen) [reply src/XGBoostSharp/lib/Booster.cs:43]: Fixed!

File: src/XGBoostSharp/lib/Booster.cs [hunk 4]
```diff
@@ -39,6 +40,40 @@ public Booster(string fileName)
...
+        finally
+        {
+            if (handle != IntPtr.Zero)
+            {
+                Marshal.FreeHGlobal(handle);
+            }
+        }
+    }
+
+    public byte[] SaveRaw(ModelFormat rawFormat = ModelFormat.Ubj)
```
[the remark this answers is not in the document]
Comment (madskonradsen) [reply src/XGBoostSharp/lib/Booster.cs]: Yes, makes sense :)

**Comment by madskonradsen** [comment]
Fixed the PR comments. Makes sense to have the SafeBufferHandle :) I also just dawned on me that the DMatrix only takes floats and not the other features_types like categorial yet https://github.com/dmlc/xgboost/blob/3f9bfaf86e6db6a4f54734aa7d164df55aa69ef6/python-package/xgboost/core.py#L855 :D But that is of course also what you said, that it's a starting point :)

**Comment by mdabros** [comment]
> Fixed the PR comments. Makes sense to have the SafeBufferHandle :) I also just dawned on me that the DMatrix only takes floats and not the other features_types like categorial yet https://github.com/dmlc/xgboost/blob/3f9bfaf86e6db6a4f54734aa7d164df55aa69ef6/python-package/xgboost/core.py#L855 :D But that is of course also what you said, that it's a starting point :)

@madskonradsen True, and yes, there are many things to add before we have something that can be considered somewhat feature complete :-)

**Review by mdabros — APPROVED** [review]
@madskonradsen Changes looks good, so I will merge 👍.  I will try to look into using safe handle directly for the native methods some time this week, and then release a new version.

**Pull request merged**